### ISSUE TITLE
esmf@8.3.0b09 pio variant error after recent update

### ISSUE THREAD

**Issue by climbfuji** [JCSDA/spack-stack#312]
**Describe the bug**
After yesterday's merge of the ESMF update, we can't install esmf@8.3.0b09 with pio anymore.

Version esmf@8.3.0b09 still uses pio the "old" way (i.e. it uses its own, old internal version of pio), but the way the variants are encoded make spack treat esmf@8.3.0b09 same as esmf@8.3.0 (a recurring problem with the esmf beta snapshot naming schemes), i.e. it expects an option parallelio to enable the parallelio interface using an external (=spack) parallelio installation.

**To Reproduce**
Tro to install esmf@8.3.0b09+pio or esmf@8.3.0b09+parallelio - neither will work.

**Expected behavior**
esmf@8.3.0b09+pio installs that version of ESMF with internal parallelio.

**System:**
Any system.

**Additional context**
I have a pretty nasty workaround for this problem:
```
diff --git a/var/spack/repos/builtin/packages/esmf/package.py b/var/spack/repos/builtin/packages/esmf/package.py
index 814384cd73..7a9ff43ad1 100644
--- a/var/spack/repos/builtin/packages/esmf/package.py
+++ b/var/spack/repos/builtin/packages/esmf/package.py
@@ -303,7 +303,13 @@ def edit(self, spec, prefix):
         ##############
         # ParallelIO #
         ##############
-        if "+parallelio" in spec and "+mpi" in spec:
+        # Nasty workaround for version 8.3.0b09 ...
+        if "@8.3.0b09" in spec:
+            if "+parallelio" in spec:
+                os.environ["ESMF_PIO"] = "internal"
+            else:
+                os.environ["ESMF_PIO"] = "OFF"
+        elif "+parallelio" in spec and "+mpi" in spec:
             os.environ["ESMF_PIO"] = "external"
             os.environ["ESMF_PIO_LIBPATH"] = spec["parallelio"].prefix.lib
             os.environ["ESMF_PIO_INCLUDE"] = spec["parallelio"].prefix.include
```
This way, one can install esmf@8.3.0b09+parallelio (using an internal parallelio - misleading and confusing, but it works).

**Comment by climbfuji** [comment]
@jedwards4b FYI

**Comment by jedwards4b** [comment]
@climbfuji  I think that this is the case for 8.3.0b01 thru 8.3.0b12.    I can work on something similar to what you've done but which will handle all of these beta tags.

**Comment by climbfuji** [comment]
> @climbfuji I think that this is the case for 8.3.0b01 thru 8.3.0b12. I can work on something similar to what you've done but which will handle all of these beta tags.

That's great, thank you. At some point we want to get rid of those beta snapshots in the spack packages, I suppose, and use the official releases. But right now the ufs-weather-model still requires 8.3.0b09 unless things have changed and I didn't notice.

**Comment by climbfuji** [comment]
I also suggest you add yourself to the maintainers of the package, if you want.

**Comment by jedwards4b** [comment]
I wasn't able to construct a fix that works for all 8.3.0beta tags but since b09 is the only one explicitly mentioned maybe that's okay.   Here is a possibly cleaner fix for the issue.   
```
diff --git a/var/spack/repos/builtin/packages/esmf/package.py b/var/spack/repos/builtin/packages/esmf/package.py
index 814384c..8215a10 100644
--- a/var/spack/repos/builtin/packages/esmf/package.py
+++ b/var/spack/repos/builtin/packages/esmf/package.py
@@ -19,7 +19,7 @@ class Esmf(MakefilePackage):
     url = "https://github.com/esmf-org/esmf/archive/ESMF_8_0_1.tar.gz"
     git = "https://github.com/esmf-org/esmf.git"
 
-    maintainers = ["climbfuji"]
+    maintainers = ["climbfuji", "jedwards4b"]
 
     # Develop is a special name for spack and is always considered the newest version
     version("develop", branch="develop")
@@ -58,6 +58,7 @@ class Esmf(MakefilePackage):
     variant("netcdf", default=True, description="Build with NetCDF support")
     variant("pnetcdf", default=True, description="Build with pNetCDF support")
     variant("xerces", default=True, description="Build with Xerces support")
+
     variant(
         "parallelio",
         default=True,
@@ -65,11 +66,23 @@ class Esmf(MakefilePackage):
         when="@8.3:",
     )
     variant(
+        "parallelio",
+        default=False,
+        description="Build with external parallelio library",
+        when="@8.3.b09",
+    )
+    variant(
         "pio",
         default=True,
         description="Enable Internal ParallelIO support",
         when="@:8.2.99",
     )
+    variant(
+        "pio",
+        default=True,
+        description="Enable Internal ParallelIO support",
+        when="@8.3.0b09",
+    )
     variant("debug", default=False, description="Make a debuggable version of the library")
 
     # Required dependencies
```

**Comment by jedwards4b** [comment]
@climbfuji What is preventing UFS using the 8.3.0 release code?

**Comment by climbfuji** [comment]
> @climbfuji What is preventing UFS using the 8.3.0 release code?

This is a "redacted" version of an email summarizing the problem that I received on June 13 from NOAA-EMC. I don't know if this has been fixed and if yes, in which ESMF version/beta snapshot.

@DusanJovic-NOAA Can you comment?

> FYI. We are looking into the ESMF 830 issue in the fv3atm native grid output from ESMF function calls in UFS WM.
> ...
> I looked at why the attribute type is not passed down to PIO/netcdf correctly from the esmf fields and it looks to me when ESMCI::Info is copied to a new info object (in Info::get function), the type storage is not properly initialized
> ...
> Can you apply this patch, recompile esmf and try to run the failed test in ufs-weather-model
```
$ git diff 
diff --git a/src/Infrastructure/Base/src/ESMCI_Info.C b/src/Infrastructure/Base/src/ESMCI_Info.C 
index 8f9d996f07..38915aebb6 100644 
--- a/src/Infrastructure/Base/src/ESMCI_Info.C 
+++ b/src/Infrastructure/Base/src/ESMCI_Info.C 
@@ -858,8 +858,11 @@ void Info::get(ESMCI::Info &info, key_t &key) const { 
#endif 
 
  json j; 
+  json const *ts = nullptr; 
  try { 
    j = this->get<json>(key); 
+    json::json_pointer jpath = this->formatKey(key); 
+    update_json_pointer(this->getTypeStorage(), &ts, jpath, true); 
 
#if 0 
    std::string msg2 = std::string(ESMC_METHOD) + ": j dump=" + j.dump(); 
@@ -867,9 +870,11 @@ void Info::get(ESMCI::Info &info, key_t &key) const { 
#endif 
     
    check_init_from_json(j); 
+    check_init_from_json(*ts); 
  } 
  ESMF_CATCH_INFO 
  info.getStorageRefWritable() = std::move(j); 
+  info.getTypeStorageWritable() = std::move(*ts); 
 
#if 0 
  std::string msg3 = std::string(ESMC_METHOD) + ": info dump=" + info.dump(0);
```

**Comment by climbfuji** [comment]
> I wasn't able to construct a fix that works for all 8.3.0beta tags but since b09 is the only one explicitly mentioned maybe that's okay. Here is a possibly cleaner fix for the issue.
> 
> ```
> diff --git a/var/spack/repos/builtin/packages/esmf/package.py b/var/spack/repos/builtin/packages/esmf/package.py
> index 814384c..8215a10 100644
> --- a/var/spack/repos/builtin/packages/esmf/package.py
> +++ b/var/spack/repos/builtin/packages/esmf/package.py
> @@ -19,7 +19,7 @@ class Esmf(MakefilePackage):
>      url = "https://github.com/esmf-org/esmf/archive/ESMF_8_0_1.tar.gz"
>      git = "https://github.com/esmf-org/esmf.git"
>  
> -    maintainers = ["climbfuji"]
> +    maintainers = ["climbfuji", "jedwards4b"]
>  
>      # Develop is a special name for spack and is always considered the newest version
>      version("develop", branch="develop")
> @@ -58,6 +58,7 @@ class Esmf(MakefilePackage):
>      variant("netcdf", default=True, description="Build with NetCDF support")
>      variant("pnetcdf", default=True, description="Build with pNetCDF support")
>      variant("xerces", default=True, description="Build with Xerces support")
> +
>      variant(
>          "parallelio",
>          default=True,
> @@ -65,11 +66,23 @@ class Esmf(MakefilePackage):
>          when="@8.3:",
>      )
>      variant(
> +        "parallelio",
> +        default=False,
> +        description="Build with external parallelio library",
> +        when="@8.3.b09",
> +    )
> +    variant(
>          "pio",
>          default=True,
>          description="Enable Internal ParallelIO support",
>          when="@:8.2.99",
>      )
> +    variant(
> +        "pio",
> +        default=True,
> +        description="Enable Internal ParallelIO support",
> +        when="@8.3.0b09",
> +    )
>      variant("debug", default=False, description="Make a debuggable version of the library")
>  
>      # Required dependencies
> ```

This is a much nicer fix than mine.
